ucsmsdk/utils/ucstechsupport.py

```diff
@@ -19,11 +19,258 @@
+    ts_cmd_opt.major_opt_type = TechSupOptsConsts.MAJOR_OPT_TYPE_UCSM_MGMT
+
+
+def _validate_chassis_options(kwargs):
+    if not _is_valid_arg("chassis_id", kwargs):
+        raise UcsValidationException('chassis_id should be specified')
+    if "cimc_id" not in kwargs and "iom_id" not in kwargs:
+        raise UcsValidationException('cimc_id/iom_id should be specified')
+
+
+def _set_ts_options_chassis(ts_cmd_opt, kwargs):
+    _validate_chassis_options(kwargs)
+    ts_cmd_opt.chassis_id = str(kwargs["chassis_id"])
+    ts_cmd_opt.major_opt_type = TechSupOptsConsts.MAJOR_OPT_TYPE_CHASSIS
+    if _is_valid_arg("cimc_id", kwargs):
+        ts_cmd_opt.chassis_cimc_id = str(kwargs["cimc_id"])
+
+        if _is_valid_arg("adapter_id", kwargs):
+            ts_cmd_opt.cimc_adapter_id = str(kwargs["adapter_id"])
+        else:
+            ts_cmd_opt.cimc_adapter_id = TechSupOptsConsts.CIMC_ADAPTER_ID_ALL
+    elif _is_valid_arg("iom_id", kwargs):
+        ts_cmd_opt.chassis_iom_id = str(kwargs["iom_id"])
+
+
+def _validate_rackserver_options(kwargs):
+    if not _is_valid_arg("rack_server_id", kwargs):
+        raise UcsValidationException('rack_server_id should be specified')
+
+
+def _set_ts_options_rackserver(ts_cmd_opt, kwargs):
+    _validate_rackserver_options(kwargs)
+    ts_cmd_opt.rack_server_id = str(kwargs["rack_server_id"])
+
+    if _is_valid_arg("rack_adapter_id", kwargs):
+        ts_cmd_opt.rack_server_adapter_id = str(kwargs["rack_adapter_id"])
+        ts_cmd_opt.major_opt_type = TechSupOptsConsts.MAJOR_OPT_TYPE_SERVER
+    else:
+        ts_cmd_opt.rack_server_adapter_id = TechSupOptsConsts.RACK_SERVER_ADAPTER_ID_ALL
+
+
+def _validate_fex_options(kwargs):
+    if not _is_valid_arg("fex_id", kwargs):
+        raise UcsValidationException('fex_id should be specified')
+
+
+def _set_ts_options_fex(ts_cmd_opt, kwargs):
+    _validate_fex_options(kwargs)
+    ts_cmd_opt.fab_ext_id = str(kwargs["fex_id"])
+    ts_cmd_opt.major_opt_type = TechSupOptsConsts.MAJOR_OPT_TYPE_FEX
+
+
+def _set_ts_options(option, ts_cmd_opt, kwargs):
+    if option == "ucsm":
+        _set_ts_options_ucsm(ts_cmd_opt, kwargs)
+    elif option == "ucsm-mgmt":
+        _set_ts_options_ucsm_mgmt(ts_cmd_opt, kwargs)
+    elif option == "chassis":
+        _set_ts_options_chassis(ts_cmd_opt, kwargs)
+    elif option == "fabric-extender":
+        _set_ts_options_fex(ts_cmd_opt, kwargs)
+    elif option == "rack-server":
+        _set_ts_options_rackserver(ts_cmd_opt, kwargs)
+    else:
+        raise UcsValidationException('Unrecognised option value: ' + option)
+
+
+def poll_wait_for_tech_support(handle, ts_mo, timeout):
+    from ..mometa.sysdebug.SysdebugTechSupport import \
+        SysdebugTechSupportConsts as TechSupConsts
+
+    poll_interval = 5
+    duration = timeout
+
+    while True:
+        try:
+            ts = handle.query_dn(ts_mo.dn)
+            if ts.oper_state == TechSupConsts.OPER_STATE_AVAILABLE:
```

Comment:
Can we also check for `OPER_STATE_UNAVAILABLE` or `OPER_STATE_FAILED` and throw an error early instead of timing out ?